### DISCUSSION TITLE
How I can get a theme color and apply to a style?

### DISCUSSION THREAD

**Discussion by Andre-LA** [orbitalquark/textadept#190] [Q&A] [answered]
Hello, thanks for making such useful editor, I really like it! :]

I'm trying to apply a background using theme colors, but it always uses a black background, my code (I'm trying to use `keyword` just for testing):

```lua
lex:add_style('preprocessor_token', lexer.STYLE_PREPROCESSOR .. {bold = true, back = lexer.styles.keyword})
```

Using Textadept 11.3

Example:

![image](https://user-images.githubusercontent.com/8538122/159594963-f5bf1b73-2046-481b-99d2-c19ff2817e21.png)

For instance, it works if I use the one of the base16 colors:

![image](https://user-images.githubusercontent.com/8538122/159595261-a5753faf-0fa4-4249-b6e0-8d3e381cdb60.png)

Looking at the base16 module, I noticed that [it contains a `fore` field](https://github.com/rgieseke/base16-textadept/blob/d8a4061ed28916e2a66230d2d9de620fb63ec1b2/themes/base16-apathy.lua#L53)

But trying to access `fore` also doesn't works, but it's just invisible instead of black:

![image](https://user-images.githubusercontent.com/8538122/159595485-0498b790-1593-4852-b507-4a7a68efffa7.png)

---

Note: I'm trying to mimic [nelua-sublime's yellow background](https://github.com/edubart/nelua-sublime) for preprocessor context.

**Comment by snoopy** [comment]
It looks like you are trying to define colors in the lexer itself which is probably not going to work. Instead set the background color you want in your theme file.  
Take this style for example:
`styles.preprocessor = {fore = colors.functions}`  
To add a background color simply add it:
`styles.preprocessor = {fore = colors.functions, back = colors.label}`

**Comment by orbitalquark** [comment, marked as answer]
As noted in the documentation[1], you generally should not specify colors in lexers. If you do, you need to specify an actual color from `lexer.colors`, not a derived color from a style (e.g. `lexer.styles.*.fore` will not work). This is due to the internal implementation of style references and not something that can easily be explained. As an example, the 'diff' lexer defines styles with colors[2].

[1]: https://orbitalquark.github.io/textadept/api.html#styles-and-styling
[2]: https://github.com/orbitalquark/scintillua/blob/530810c22ab92b2dbc0e0e627a5b9de3cd236475/lexers/diff.lua#L21

**Comment by Andre-LA** [comment]
Understood, I'll focus on other areas on the lexer and drop this ideia then.

Thanks for the replies!
